apps/aehttp/src/aehttp_cache_expires.erl

```diff
@@ -0,0 +1,218 @@
+%% Implements expires/3 similar to cowboy_rest expires/2 callback
+%% customized with additional OperationId (first) argument for easy switching
+%%
+%% Get the top block (either key or micro block)
+%% Cannot be cached for more than MICRO_BLOCK_CYCLE (3s)
+%%
+%% Get a micro block
+%% Can be cached aggressively, however micro-blocks of last X key blocks can be reorganized by forks,
+%% thus their cache time should not exceed EXPECTED_BLOCK_MINE_RATE_MINUTES (3 min.)
+%% Expires = ExpectedNextBlockTime() -> TopBlockTime + InterBlockTime
+%%
+%% Also micro blocks can be reorganized because of micro-forks,
+%% so micro-blocks of current generation cannot be cached for more than MICRO_BLOCK_CYCLE (3s)
+%% Expires = ExpectedNextMicroBlockTime() -> TopMicroBlockTime + InterMicroBlockTime
+%%
+%% Get a key block
```

Comment:
this one is the most controversial. We have key blocks happening after 30sec or after 7 minutes...

Reply:
Wouldn't it be better to listen for new block events and invalidate the cache based on this? Not sure we have exactly the right events, but that could be fixed I think?!

Reply:
Exactly the only thing we can make. Price to pay to have a cache

Reply:
I think there is misunderstanding here (I updated the PR description). This is about HTTP cache headers that's a spec. for HTTP caching implementation in the clients. There is no caching in the node.

The only consequence here is that a block can be cached for a bit longer than it's real generation time if the block is found earlier than expected. However, that time won't exceed the expected block time. So on average, the clients should have their fresh copy on time, it small exceptions.

I don't think this is too different from the protocol expectations and implementation ? 🤔

Reply:
Yes our comments are invalid. The problem is the guidance for clients when to consider resource expired and we can't do anything about it from server side.

Reply:
Maybe use the most pessimistic one? (like the shortest keyblock time over the last day?)

Reply:
Ah, client caching 🤦‍♂